### PR TITLE
Handle NXDOMAIN, REFUSED and log errors

### PR DESCRIPTION
**Pull request by thaJeztah**
relates to https://github.com/moby/moby/pull/37180, https://github.com/moby/moby/issues/20494, https://github.com/docker/compose/issues/5991

- NXDOMAIN is an authoritive answer, so when receiving an NXDOMAIN, we're done.
  From RFC 1035: Name Error - Meaningful only for responses from an authoritative
  name server, this code signifies that the domain name referenced in the query
  does not exist.
  FROM RFC 8020: When an iterative caching DNS resolver receives an NXDOMAIN
  response, it SHOULD store it in its cache and then all names and resource
  record sets (RRsets) at or below that node SHOULD be considered unreachable.
  Subsequent queries for such names SHOULD elicit an NXDOMAIN response.
- REFUSED can be a transitional status: (https://www.ietf.org/rfc/rfc1035.txt)
  The name server refuses to perform the specified operation for
  policy reasons.  For example, a name server may not wish to provide the
  information to the particular requester, or a name server may not wish to
  perform a particular operation (e.g., zone)

Other errors are now logged as debug-message, which can be useful for troubleshooting.

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @ddebroy @ctelfer PTAL; t.b.h., handling of `REFUSED` is not well defined in the specs, but I found references to it also being used during periods when (e.g.) the DNS's cache is being refreshed

**Comment by Sytten** [comment]
`nslookup` seems to handle the `REFUSED` as a `NXDOMAIN` and in both cases it goes to the next DNS server.

File: resolver.go [hunk 1]
```diff
@@ -490,35 +490,51 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
...
+			case dns.RcodeNameError:
+				// Server returned NXDOMAIN. Stop resolution: it's an authoritative answer (see RFC 8020: https://tools.ietf.org/html/rfc8020#section-2)
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), name)
+				break
+			case dns.RcodeSuccess:
+				// All is well
+			default:
+				// TODO should we log as error, and break here?
+				// Server gave some error. Log the error, but proceed as usual
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s (code %d) for %q", proto, extDNS.IPStr, statusString(resp.Rcode), resp.Rcode, name)
```
Comment (Sytten) [inline resolver.go]: we should probably handle all codes to match the RFC.

**Comment by thaJeztah** [comment]
One _possible_ issue is the part that "`NXDOMAIN` is "Meaningful only for responses from an authoritative name server", which could be interpreted as "any other name server returning this error should be ignored"

Reasons for _not_ continuing. Admitted; I think these are about [_recursive_ DNS, not _iterating_ DNS](https://www.slashroot.in/difference-between-iterative-and-recursive-dns-query) setups, so possibly interpreting it incorrect; [Understanding NXDOMAIN Attack Methods](https://community.infoblox.com/t5/Company-Blog/Understanding-NXDOMAIN-Attack-Methods/ba-p/3949), [Mitigating Possible NXDOMAIN Attacks](https://docs.infoblox.com/display/NAG8/Mitigating+Possible+NXDOMAIN+Attacks)

**Comment by Sytten** [comment]
I agree on the part of the authoritative name server. I am not that familiar with DNS, is there a way to know by the response if the name server is an authority?
I understand the security concern, we should probably give the user the choice to enable/disable this particular continuation for NXDOMAIN. I can see both use cases being used.

**Comment by ctelfer** [comment]
Isn't whether the server is considered authoritative specified by a bit in the header?
```
RFC 1035        Domain Implementation and Specification    November 1987


4.1.1. Header section format

The header contains the following fields:

                                    1  1  1  1  1  1
      0  1  2  3  4  5  6  7  8  9  0  1  2  3  4  5
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                      ID                       |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |QR|   Opcode  |AA|TC|RD|RA|   Z    |   RCODE   |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                    QDCOUNT                    |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                    ANCOUNT                    |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                    NSCOUNT                    |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                    ARCOUNT                    |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
...
AA              Authoritative Answer - this bit is valid in responses,
                and specifies that the responding name server is an
                authority for the domain name in question section.

                Note that the contents of the answer section may have
                multiple owner names because of aliases.  The AA bit
                corresponds to the name which matches the query name, or
                the first owner name in the answer section.
```

Not sure how this is exposed in our DNS library though.  Will check.

**Comment by thaJeztah** [comment]
There may be a field for that yes that we can access through that library

**Comment by ctelfer** [comment]
Looks like it should be `resp.Authoritative` to test whether that bit was set in the header.

File: resolver.go [hunk 2]
```diff
@@ -490,35 +490,53 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
...
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), name)
+				continue
+			case dns.RcodeRefused:
+				// Server returned REFUSED: this can be a transitional status, so continue with the next external DNS server
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), name)
+				continue
+			case dns.RcodeNameError:
+				// Server returned NXDOMAIN. Stop resolution if it's an authoritative answer (see RFC 8020: https://tools.ietf.org/html/rfc8020#section-2)
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), name)
+				if resp.Authoritative {
```
Comment (thaJeztah) [inline resolver.go]: Added a check if the response is Authoritative

**Comment by jeff-jk** [comment]
Does this need another set of eyes to review or are you guys considering this complete? I'm still seeing  TODO in the changes so it didn't seem to be entirely finished yet.

**Comment by thaJeztah** [comment]
The TODO is around code that keeps the current behavior (w.r.t. handling errors), which is to _log_ the error, but continue processing. I didn't want to change that in this PR (to not combine multiple changes in behavior), but open to input from the maintainers @ctelfer @fcrisciani

File: resolver.go (hunk 1 above)
Comment (ctelfer) [reply resolver.go]: I notice above that the branches for `RcodeServerFailure` and `RcodeRefused` are identical except for the comments.  I guess that's OK since they call out specific rationale for the decision, but they could be collapsed into the `default:` branch w/ no loss of information.  (just an extra code # in the log).

So, then took a look at https://tools.ietf.org/html/rfc6895 to see what codes were missing and there are a bunch.   For most of them, I think the safe thing to do would just be log and `continue` on to the next server's query.  So this would mean putting `continue` in the `default` branch above.  Should be safe, for example, to go on to the next query when we get `FormErr`, `NotImp` (should never get either of these), `Refused`, (another server might not refuse)  and `NotAuth` (another server might be authoritative).  The codes `YXDomain`, `YXRRSet`, `NXRRSet`, `NotZone` seem to indicate that the query is for a bad name.  If so, `continue`ing is safe in a sense but doesn't return good error info to the container.  We should never get any of the codes > 16 because we aren't doing DNSSEC, but if we did, we certainly shouldn't accept anything in the answer section from them.

TLDR:
  * the default branch should at least `continue` so we don't accept answers w/ potential error codes
  * we could collapse `RcodeRefused` and `RcodeServerFailure` into said branch but can see leaving it explicit
  * more robust handling would return proper responses for some of these errors but that could be left for another PR

**Comment by Sytten** [comment]
Any progress on that?

**Comment by ctelfer** [comment]
FWIW, other than needing a `continue` for unknown codes (see above -- default branch of the switch statement), it LGTM.  The other issues raised could go in another PR for another day.

**Comment by thaJeztah** [comment]
@ctelfer sorry for the delay; forgot about this one. I updated the PR with the following changes;

- combined the `RcodeServerFailure` and `RcodeRefused` in a single "case"
- added `continue` for unknown errors
- also added `continue` for *non-authoritative* `NXDOMAIN` (assuming an `NXDOMAIN` is an "error", and in case it's non-authoritative we can continue with the next DNS) 

PTAL
